File: 240313/트로미노/tromino.cpp
#include <iostream>
#include <algorithm>
using namespace std;
#define MAX_NM 200

int n, m;
int arr[MAX_NM][MAX_NM];
int dx[4] = {-1, 0, 1, 0}, dy[4] = {0, 1, 0, -1};

bool InRange(int x, int y) {
    return 0<=x && x<n && 0<=y && y<m;
}

int main() {
    // 여기에 코드를 작성해주세요.
    cin >> n>>m;
    for(int i=0; i<n; i++) {
        for(int j=0; j<m; j++) {
            cin >> arr[i][j];
        }
    }

    // 첫번쨰 종류의 블럭에서 가장 큰 수 구하기
    int ans = 0;
    for(int i=0; i<n; i++) {
        for(int j=0; j<m; j++) {
            for(int k=0; k<4; k++) {
                int dir_num = k;
                int sum = arr[i][j];

                int nx1 = i+dx[dir_num], ny1 = j+dy[dir_num];
                dir_num = (dir_num+1)%4;
                int nx2 = i+dx[dir_num], ny2 = j+dy[dir_num];
                if(!InRange(nx1, ny1) || !InRange(nx2, ny2)) continue;

                sum = sum+arr[nx1][ny1]+arr[nx2][ny2];
                ans = max(ans, sum);
            }
        }
    }

    // 두번째 종류의 블럭에서 가장 큰 수 구하기
    for(int i=0; i<n; i++) {
        for(int j=0; j<m; j++) {
            for(int k=0; k<2; k++) {
                int dir_num = k;
                int sum = arr[i][j];

                int nx1 = i+dx[dir_num], ny1 = j+dy[dir_num];
                dir_num = (dir_num+2)%4;
                int nx2 = i+dx[dir_num], ny2 = j+dy[dir_num];
                if(!InRange(nx1, ny1) || !InRange(nx2, ny2)) continue;

                sum = sum+arr[nx1][ny1]+arr[nx2][ny2];
                ans = max(ans, sum);
            }
        }
    }

    cout << ans;
    return 0;
}

File: 241227/트로미노/tromino.py
n, m = tuple(map(int, input().split()))

grid = [
    list(map(int, input().split()))
    for _ in range(n)
]

dx, dy = [-1, 0, 1, 0], [0, 1, 0, -1]

def InRange(x, y):
    return 0<=x<n and 0<=y<n

def Sum1(x, y):
    summ = 0
    for d in range(4):
        nx1, ny1 = x+dx[d], y+dy[d]
        nx2, ny2 = x+dx[(d+1)%4], y+dy[(d+1)%4]

        if(InRange(nx1, ny1) and InRange(nx2, ny2)):
            summ = max(summ, grid[x][y] + grid[nx1][ny1] + grid[nx2][ny2])
    
    return summ

def Sum2(x, y):
    summ = 0
    for d in range(2):
        nx1, ny1 = x+dx[d], y+dy[d]
        nx2, ny2 = x+dx[d+2], y+dy[d+2]

        if(InRange(nx1, ny1) and InRange(nx2, ny2)):
            summ = max(summ, grid[x][y] + grid[nx1][ny1] + grid[nx2][ny2])
    
    return summ

ans = 0

for i in range(n):
    for j in range(m):
        ans = max(ans, Sum1(i, j))
        ans = max(ans, Sum2(i, j))

print(ans)